iroha_p2p/src/peer.rs

```diff
@@ -0,0 +1,325 @@
+use std::io::{Error, ErrorKind};
+
+use async_stream::stream;
+use futures::Stream;
+use rand::{Rng, RngCore};
+use tokio::io;
+use tokio::io::{AsyncReadExt, AsyncWriteExt};
+use tokio::net::tcp::{OwnedReadHalf, OwnedWriteHalf};
+use tokio::net::TcpStream;
+use ursa::kex::KeyExchangeScheme;
+use ursa::kex::x25519::X25519Sha256;
+use ursa::keys::{PrivateKey, PublicKey};
+
+use iroha_actor::{Actor, Addr, Context, Handler};
+use iroha_data_model::prelude::PeerId;
+use iroha_logger::warn;
+
+use crate::message::{Message, MessageResult};
+use crate::network::{ReceivedPeerMsg, SendPeerMsg};
+use crate::Network;
+use ursa::encryption::symm::{SymmetricEncryptor, Encryptor};
+
+const MAX_MESSAGE_LENGTH: usize = 2 * 1024 * 1024;
+const MAX_HANDSHAKE_LENGTH: usize = 256;
+pub const DEFAULT_AAD: &[u8; 12] = b"Iroha2Iroha2";
+
+pub struct Peer<E: Encryptor + Send + 'static> {
+    id: PeerId,
+    read: Option<OwnedReadHalf>,
+    write: Option<OwnedWriteHalf>,
+    state: State,
+    secret_key: PrivateKey,
+    public_key: PublicKey,
+    cipher: Option<SymmetricEncryptor<E>>,
```

Comment:
You will always have encryption. (I mean, it can be mock, but still

Reply:
We create Encryptor only after successful handshake. Until then it will be None.

Reply:
Makes sence